botorch/optim/fit.py

```diff
@@ -179,119 +183,80 @@ def store_iteration(xk):
+            determistic optimizers). Enabling this is only recommended when
+            working with large training data sets (n>2000).
 
     Returns:
-        2-element tuple containing
-        - mll with parameters optimized in-place.
-        - Dictionary with the following key/values:
-        "fopt": Best mll value.
-        "wall_time": Wall time of fitting.
-        "iterations": List of OptimizationIteration objects with information on each
-        iteration. If track_iterations is False, will be empty.
-
-    Example:
-        >>> gp = SingleTaskGP(train_X, train_Y)
-        >>> mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
-        >>> mll.train()
-        >>> fit_gpytorch_torch(mll)
-        >>> mll.eval()
+        The provided `mll` instance along with a dictionary containing the final
+        value returned by `closure` and overall wall time of the GD.
     """
-    optim_options = {"maxiter": 100, "disp": True, "lr": 0.05}
-    optim_options.update(options or {})
-    exclude = optim_options.pop("exclude", None)
-    if exclude is None:
-        mll_params = list(mll.parameters())
-    else:
-        mll_params = [
-            v for k, v in filter(create_name_filter(exclude), mll.named_parameters())
-        ]
-
-    optimizer = optimizer_cls(
-        params=[{"params": mll_params}],
-        **_filter_kwargs(optimizer_cls, **optim_options),
-    )
+    if stopping_criterion is None and step_limit is None or step_limit == float("inf"):
+        raise RuntimeError("No termination conditions were detected.")
 
-    # get bounds specified in model (if any)
-    bounds_: ParameterBounds = {}
-    if hasattr(mll, "named_parameters_and_constraints"):
-        for param_name, _, constraint in mll.named_parameters_and_constraints():
-            if constraint is not None and not constraint.enforced:
-                bounds_[param_name] = constraint.lower_bound, constraint.upper_bound
+    start_time = monotonic()
+    if closure is None:
+        closure = get_loss_closure(mll)
 
-    # update with user-supplied bounds (overwrites if already exists)
+    param_dict, bounds_dict = get_parameters_and_bounds(mll, requires_grad=True)
     if bounds is not None:
-        bounds_.update(bounds)
+        bounds_dict.update(bounds)
 
-    iterations = []
-    t1 = monotonic()
+    if stopping_criterion is DEFAULT:
+        stopping_criterion = ExpMAStoppingCriterion()
 
-    param_trajectory: Dict[str, List[Tensor]] = {
-        name: [] for name, param in mll.named_parameters()
-    }
-    loss_trajectory: List[float] = []
-    i = 0
-    stop = False
-    stopping_criterion = ExpMAStoppingCriterion(
-        **_filter_kwargs(ExpMAStoppingCriterion, **optim_options)
-    )
-    train_inputs, train_targets = mll.model.train_inputs, mll.model.train_targets
-    while not stop:
-        optimizer.zero_grad()
+    if isinstance(torch_optimizer, (type, partial)):
+        torch_optimizer = torch_optimizer(params=list(param_dict.values()))
```

Comment:
I am not the biggest fan of the pattern of allowing this to be either an instance or a factory function. Not sure if there is an elegant alternative though.

Reply:
Would factory only be better or worse? Note that instances can still be passed as `lambda **_: instance`, but the typing is now more explicit.

Reply:
My $0.02: being able to pass an Instance is nicer than the factory, if only because then I don't necessarily need all the parameters I want to optimize to live somewhere on `mll`. There's even a reasonable argument to be made for accepting lists of Optimizers and stepping all of them, to support e.g. NGD or different Optimizers / learning rates for the GP vs a DNN

Reply:
> being able to pass an Instance is nicer than the factory, if only because then I don't necessarily need all the parameters I want to optimize to live somewhere on mll.

So, here's what we can do. We can split up `fit_gpytorch_torch` into a pair of methods: i) a generic torch-based gradient descent method and ii) a wrapper that provides mll-based model fitting conveniences. This seems to address your first point.

> There's even a reasonable argument to be made for accepting lists of Optimizers and stepping all of them, to support e.g. NGD or different Optimizers / learning rates for the GP vs a DNN

Interesting idea. I like that this eliminates the need for `step_limit`. If you wanted to go further, you could pass in a list of `OptimizationStep` objects with slots for `closure`, `optimizer`, etc . Unclear to me whether we run into any scheduling issues here.